### PR TITLE
Transformation Mappings Read and Create

### PR DESCRIPTION
**Pull request by jntullo**
Adds reads and creates of Transformation Mappings to the API:

Sample create:
```
POST /api/transformation_mappings
{"name": "new transformation mapping",
 "description": "clusters to clusters",
 "state": "draft",
 "transformation_mapping_items": [
     { "source": "/api/clusters/10", "destination": "/api/clusters/1" },
     { "source": "/api/clusters/11", "destination": "/api/clusters/1" },
     { "source": "/api/date_stores/12", "destination": "/api/data_stores/13" },
     { "source": "/api/data_stores/2", "destination": "/api/data_stores/13" },
     { "source": "/api/lans/2", "destination": "/api/lans/13" } 
     { "source": "/api/lans/3", "destination": "/api/lans/13" } 
   ]
}
```

Create will accept `transformation_mapping_items` as part of the request with href references to the source and destination. 

A call to
```
GET /api/transformation_mappings
```
Will return all of the transformation mappings.

```
GET /api/transformation_mappings/:id
```
Will return a particular transformation mapping.

Transformation Mapping Items may be retrieved via:
```
GET /api/transformation_mappings/:id?attributes=transformation_mapping_items
```
which will return
```
{  
   "href":"http://localhost:3000/api/transformation_mappings/10000000000001",
   "id":"10000000000001",
   "name":"Cluster and Datastores",
   "description":"Cluster and Datastores Mapping",
   "comments":null,
   "state":"draft",
   "options":{  

   },
   "tenant_id":null,
   "created_at":"2018-02-07T16:02:39Z",
   "updated_at":"2018-02-07T16:02:39Z",
   "transformation_mapping_items":[  
      {  
         "id":"10000000000001",
         "source_id":"10000000000002",
         "source_type":"EmsCluster",
         "destination_id":"10000000000005",
         "destination_type":"EmsCluster",
         "transformation_mapping_id":"10000000000001",
         "options":{  

         },
         "created_at":"2018-02-07T16:01:44Z",
         "updated_at":"2018-02-07T16:02:39Z"
      },
      {  
         "id":"10000000000002",
         "source_id":"10000000000001",
         "source_type":"Storage",
         "destination_id":"10000000000018",
         "destination_type":"Storage",
         "transformation_mapping_id":"10000000000001",
         "options":{  

         },
         "created_at":"2018-02-07T16:05:04Z",
         "updated_at":"2018-02-07T16:05:04Z"
      }
   ]
}
```

Dependent upon:
https://github.com/ManageIQ/manageiq/pull/16787
https://github.com/ManageIQ/manageiq/pull/16947
~~and needs a factory~~

cc: @bzwei @gmcculloug

### CODE REVIEW

File: app/controllers/api/transformation_mappings_controller.rb [hunk 1]
```diff
@@ -0,0 +1,16 @@
+module Api
+  class TransformationMappingsController < BaseController
+    def create_resource(type, id, data = {})
+      raise BadRequestError, "Must specify mappings" unless data["mappings"]
+      data["mappings"] = data["mappings"].each_with_object({}) do |(k, v), new_mappings|
+        begin
+          klass = collection_config.klass(k)
```
Comment (bzwei) [inline app/controllers/api/transformation_mappings_controller.rb]: @jntullo does `collect_config` already map for example `cluster` to `EmsCluster`?
What about the GET - where do you do the translation?

**Comment by bzwei** [comment]
@jntullo ManageIQ/manageiq#16787 has the factory for transformation_mapping. Hope it satisfies your need.

File: app/controllers/api/transformation_mappings_controller.rb (hunk 1 above)
Comment (jntullo) [reply app/controllers/api/transformation_mappings_controller.rb]: @bzwei Yeah, it maps to `EmsCluster`, as defined in the [api.yml](https://github.com/ManageIQ/manageiq-api/blob/master/config/api.yml#L561-L568) - I forgot about the reverse translation (on `GET`), but I will add that in

**Comment by bzwei** [comment]
@jntullo the create example should be something like
```
POST /api/transformation_mappings
{"name": "new transformation mapping",
 "description": "clusters to clusters",
 "state": "draft",
 "transformation_mapping_items": [
     { "source": "/api/clusters/10", "destination": "/api/clusters/1" },
     { "source": "/api/clusters/11", "destination": "/api/clusters/1" },
     { "source": "/api/clusters/12", "destination": "/api/clusters/1" },
     { "source": "/api/clusters/2", "destination": "/api/clusters/13" } 
   ]
}
```

**Comment by bzwei** [comment]
@miq-bot add_label transformation

**Comment by bzwei** [comment]
@jntullo can you resolve the travis errors?

File: spec/requests/transformation_mappings_spec.rb [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+        expected = {
+          "error" => a_hash_including(
+            "kind"    => "bad_request",
+            "message" => /Must specify source and destination hrefs/
+          )
+        }
+        expect(response.parsed_body).to include(expected)
+        expect(response).to have_http_status(:bad_request)
+      end
+    end
```
Comment (abellotti) [inline spec/requests/transformation_mappings_spec.rb]: can you add a test for when the source or destination href is bogus, i.e. /api/invalid_collection/:id,  we should get a BadRequest on those.  Might need a minor change to fetch_mapping_resource. Thanks.

**Comment by abellotti** [comment]
Looks good @jntullo, just an additional test needed.

**Comment by abellotti** [comment]
Thanks @jntullo for updating this. LGTM!! 👍

**Comment by himdel** [comment]
gaprindashvili: depends on https://github.com/ManageIQ/manageiq/pull/16787

**Comment by simaishi** [comment]
Gaprindashvili backport details:

```
$ git log -1
commit 3b42f4a292d440f947d921d1919b2b87f9ceb666
Author: Alberto Bellotti <abellotti@users.noreply.github.com>
Date:   Tue Feb 20 09:56:45 2018 -0500

    Merge pull request #313 from jntullo/transformation_mappings
    
    Transformation Mappings Read and Create
    (cherry picked from commit 8772e526fe61c9c149ec36b3cc0466f0d29266d3)
```

**Pull request merged**